plutus-core/exe/Main.hs

```diff
@@ -412,15 +431,86 @@ runExample (ExampleOptions (ExampleSingle name)) = do
         Just ex -> PLC.render $ prettyExample ex
 
 
+---------------- Typechecking ----------------
+
+runTypecheck :: TypecheckOptions -> IO ()
+runTypecheck (TypecheckOptions inp fmt) = do
+  TypedProgram prog <- getProg Typed inp fmt
+  case PLC.runQuoteT $ do
+    types <- PLC.getStringBuiltinTypes ()
+    PLC.typecheckPipeline (PLC.TypeCheckConfig types) (void prog)
+    of
+       Left (e :: PLC.Error PLC.DefaultUni ()) -> T.putStrLn (PLC.displayPlcDef e) >> exitFailure
+       Right ty                                -> T.putStrLn (PLC.displayPlcDef ty) >> exitSuccess
+
+
+---------------- Evaluation ----------------
+
+runEval :: EvalOptions -> IO ()
+runEval (EvalOptions typing inp mode fmt printtime) =
+    case typing of
+      Typed -> do
+        TypedProgram prog <- getProg Typed inp fmt
+        let evaluate = case mode of
+                          CK  -> PLC.unsafeEvaluateCk  (PLC.getStringBuiltinMeanings @ (PLC.CkValue  PLC.DefaultUni))
+                          CEK -> PLC.unsafeEvaluateCek (PLC.getStringBuiltinMeanings @ (PLC.CekValue PLC.DefaultUni)) PLC.defaultCostModel
+            body = void . PLC.toTerm $ prog
+            _ = rnf body   -- Force evaluation of body to ensure that we're not timing parsing/deserialisation
+        start <- getCPUTime
+        case evaluate body of
+              PLC.EvaluationSuccess v ->
+                  do
+                    end <- getCPUTime
+                    T.putStrLn (PLC.displayPlcDef v)
+                    when (printtime == Timing) $ printExecutiontime start end
+                    exitSuccess
+              PLC.EvaluationFailure -> exitFailure
```

Comment:
Can we abstract this part at least?

Reply:
At, that's better.  I hadn't spotted that the type of v was the same in both cases.